timed/authentication.py

```diff
@@ -0,0 +1,109 @@
+import base64
+import functools
+import hashlib
+
+import requests
+from django.conf import settings
+from django.core.cache import cache
+from django.core.exceptions import SuspiciousOperation
+from django.utils.encoding import force_bytes
+from mozilla_django_oidc.auth import LOGGER, OIDCAuthenticationBackend
+
+
+class TimedOIDCAuthenticationBackend(OIDCAuthenticationBackend):
+    def get_introspection(self, access_token, id_token, payload):
+        """Return user details dictionary."""
+
+        basic = base64.b64encode(
+            f"{settings.OIDC_OP_INTROSPECT_CLIENT_ID}:{settings.OIDC_OP_INTROSPECT_CLIENT_SECRET}".encode(
+                "utf-8"
+            )
+        ).decode()
+        headers = {
+            "Authorization": f"Basic {basic}",
+            "Content-Type": "application/x-www-form-urlencoded",
+        }
+        response = requests.post(
+            settings.OIDC_OP_INTROSPECT_ENDPOINT,
+            verify=settings.OIDC_VERIFY_SSL,
+            headers=headers,
+            data={"token": access_token},
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def get_userinfo_or_introspection(self, access_token):
+        try:
+            claims = self.cached_request(
+                self.get_userinfo, access_token, "auth.userinfo"
+            )
+        except requests.HTTPError as e:
+            if not (
+                e.response.status_code in [401, 403] and settings.OIDC_CHECK_INTROSPECT
+            ):
+                raise e
+
+            # check introspection if userinfo fails (confidental client)
+            claims = self.cached_request(
+                self.get_introspection, access_token, "auth.introspection"
+            )
+            if "client_id" not in claims:
+                raise SuspiciousOperation("client_id not present in introspection")
+
+        return claims
+
+    def get_or_create_user(self, access_token, id_token, payload):
+        """Verify claims and return user, otherwise raise an Exception."""
+
+        claims = self.get_userinfo_or_introspection(access_token)
+
+        # claims_verified = self.verify_claims(userinfo) or self.verify_claims(introspection)
+
+        # if not claims_verified:
+        #     msg = "Claims verification failed"
+        #     raise SuspiciousOperation(msg)
```

Comment:
Should be removed or used to verify the claims.

Reply:
did that